src/plugins/plugin-pg.ts

```diff
@@ -13,35 +13,65 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-'use strict';
 
-var shimmer = require('shimmer');
+import {EventEmitter} from 'events';
+import * as shimmer from 'shimmer';
+import {Readable} from 'stream';
 
-var SUPPORTED_VERSIONS = '^6.x || ^7.x';
+import {Patch, Plugin} from '../plugin-types';
 
-module.exports = [
+import {pg_6, pg_7} from './types';
+
+// TS: Client#query also accepts a callback as a last argument, but TS cannot
+// detect this as it's a dependent type. So we don't specify it here.
+type PG7QueryArguments =
+    [{submit?: Function} & pg_7.QueryConfig]|[string]|[string, {}];
+type PG7QueryReturnValue = (pg_7.QueryConfig&({submit: Function}&EventEmitter)|
+                            pg_7.Query)|Promise<pg_7.QueryResult>;
+
+// tslint:disable-next-line:no-any
+function isSubmittable(obj: any): obj is {submit: Function} {
+  return typeof obj.submit === 'function';
+}
+
+const plugin: Plugin = [
   {
     file: 'lib/client.js',
-    versions: SUPPORTED_VERSIONS,
-    patch: function(Client, api) {
-      function queryWrap(query) {
-        return function query_trace() {
-          var span = api.createChildSpan({
-            name: 'pg-query'
-          });
-          var pgQuery = query.apply(this, arguments);
+    versions: '^6.x',
+    // TS: Client is a class name.
+    // tslint:disable-next-line:variable-name
+    patch: (Client, api) => {
+      shimmer.wrap(Client.prototype, 'query', (query) => {
+        return function query_trace(this: pg_6.Client) {
+          const span = api.createChildSpan({name: 'pg-query'});
           if (!api.isRealSpan(span)) {
-            return pgQuery;
+            return query.apply(this, arguments);
           }
-          if (api.enhancedDatabaseReportingEnabled()) {
-            span.addLabel('query', pgQuery.text);
-            if (pgQuery.values) {
-              span.addLabel('values', pgQuery.values);
+          const args: pg_6.QueryReturnValue[] =
+              Array.prototype.slice.call(arguments, 0);
+          if (args.length >= 1) {
```

Comment:
Nit. Doing the length check before `Array.prototype.slice.call()` would be better when there's no arguments.